enhancements/authentication/configuring-webhook-token-authenticators.md

```diff
@@ -0,0 +1,159 @@
+
+`webhookTokenAuthenticators` could be used for configuring a custom OAuth server which supports Kubernetes'
+token reviews (see [Webhook Token Authentication](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#webhook-token-authentication)
+for more details). Therefore, by implementing this enhancement, we would significantly simplify any
+drop-in replacements for the integrated oauth-server, which might be helpful in cases where we would
+like to delegate authentication to a different OpenShift cluster, as an example.
+
+### Goals
+
+1. Make it possible to configure an endpoint capable of performing kube token reviews
+2. Modify cluster-authentication-operator so that it configures authentication via the OpenShift OAuth
+   server by using the `webhookTokenAuthneticators` field instead of the current patching of the
+   kube-apiserver.
+
+## Proposal
+
+### Implementation
+
+#### API
+
+`authentications.config.openshift.io/v1` already contains the `spec.webhookTokenAuthenticators` field
+that expects an array of references to secrets in the `openshift-config` namespace. This field could
+be used for the purposes of configuring webhook token authenticators, where we would just expect references
+to secrets, that would have to be checked.
+
+The configuration mentioned above leaves quite a lot of freedom in hands of users as the referenced secrets
+would be quite general kubeconfigs which makes users be able to configure fields that could make the
+kube-apiserver access user-specified paths, which may not be desirable (although could be ruled out
+by validation).
+
+An alternative is to handle the configuration by introducing a new cluster-wide CRD:
+```
+apiVersion: config.openshift.io/v1
+kind: TokenAuthenticator
+spec:
+  ca: CM reference
+  server: string (valid https:// URL)
+  clientCertificate: CM reference
+  clientKey: Secret reference
+```
+
+#### Backend
+
+__cluster-authentication-operator__
+- observes the changes in the `authentication/cluster` resource (or observes new `TokenAuthenticator` object)
+  and performs validation of any references
+- it pushes any webhook authenticator that passes the validation to `openshift-config-managed/webhooktokenauthenticators`
+  secret, where `key=authenticator_name` where `authenticator_name` is either the name of the referenced secret or
+  the CR
+- if no webhook authenticators pass validation, the operator keeps the previous state
+
+__kube-apiserver-operator__
+- observes the `openshift-config-managed/webhooktokenauthenticators` secret, on a change it syncs it to the
+  openshift-kube-apiserver namespace and redeploys the kube-apiserver static pods while specifying the new
+  mounted files in the `webhookTokenAuthenticators` [config field](https://github.com/openshift/api/blob/992bc3a41fe6c4e5bcf27dc8080f9e5c114c33fc/kubecontrolplane/v1/types.go#L146).
+
+### Risks and Mitigations
+
+There are a couple of issues that root in the very fact of making it possible of
```

Comment:
this isn't because we made it configurable, this is because our oauth server has certain special features.  This is problem and responsibility of the piece replacing us.

Reply:
I added a line explaining that these should all be a responsibility of the integrator

Reply:
rephrase:  "Common gotchas when providing your own authentication webhook"